modules/renter/contractor/contracts.go

```diff
@@ -0,0 +1,395 @@
+	if proto.IsRevisionMismatch(err) {
+		// return unused outputs to wallet
+		txnBuilder.Drop()
+		// try again with the cached revision
+		c.mu.RLock()
+		cached, ok := c.cachedRevisions[contract.ID]
+		c.mu.RUnlock()
+		if !ok {
+			// nothing we can do; return original error
+			c.log.Printf("wanted to recover contract %v with host %v, but no revision was cached", contract.ID, contract.NetAddress)
+			return modules.RenterContract{}, err
+		}
+		c.log.Printf("host %v has different revision for %v; retrying with cached revision", contract.NetAddress, contract.ID)
+		contract.LastRevision = cached.Revision
+		// need to start a new transaction
+		txnBuilder = c.wallet.StartTransaction()
+		newContract, err = proto.Renew(contract, params, txnBuilder, c.tpool, c.hdb, c.tg.StopChan())
+	}
+	if err != nil {
+		txnBuilder.Drop() // return unused outputs to wallet
+		return modules.RenterContract{}, err
+	}
+
+	return newContract, nil
+}
+
+// threadedContractMaintenance checks the set of contracts that the contractor
+// has against the allownace, renewing any contracts that need to be renewed,
+// dropping contracts which are no longer worthwhile, and adding contracts if
+// there are not enough.
+func (c *Contractor) threadedContractMaintenance() {
+	// Threading protection.
+	err := c.tg.Add()
+	if err != nil {
+		return
+	}
+	defer c.tg.Done()
+	// Only one instance of this thread should be running at a time.
+	if !c.maintenanceLock.TryLock() {
+		return
+	}
+	defer c.maintenanceLock.Unlock()
+
+	// Renew any contracts that need to be renewed.
+	//
+	// TODO: Include checking if the contract needs more money.
+	//
+	// TODO: Add the IsGoodForRenew logic.
+	c.mu.Lock()
+	var renewSet []types.FileContractID
+	for _, contract := range c.onlineContracts() {
+		if c.blockHeight+c.allowance.RenewWindow >= contract.EndHeight() {
+			renewSet = append(renewSet, contract.ID)
+		}
+	}
+	c.mu.Unlock()
+	if len(renewSet) != 0 {
+		c.log.Printf("renewing %v contracts", len(renewSet))
+	}
+
+	// Figure out the end height and target sector count for the contracts being
+	// renewed.
+	//
+	// TODO: EndHeight should be global, and it should be picked based on the
+	// current period start, not based on the current height plus the allowance
+	// period.
+	c.mu.RLock()
+	endHeight := c.blockHeight + c.allowance.Period
+	max, err := maxSectors(c.allowance, c.hdb, c.tpool)
+	c.mu.RUnlock()
+	if err != nil {
+		return
+	}
+	// Only allocate half as many sectors as the max. This leaves some leeway
+	// for replacing contracts, transaction fees, etc.
+	numSectors := max / 2
+	// check that this is sufficient to store at least one sector
+	if numSectors == 0 {
+		return
+	}
+
+	// Loop through the contracts and renew them one-by-one.
+	for _, id := range renewSet {
+		// Quit / return in the event of shutdown.
+		select {
+		case <-c.tg.StopChan():
+			return
+		default:
+		}
```

Comment:
we already have a `tg.isStopped` method. It used to be exported but then we unexported it. I forget why.